crawler/main.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"golang.org/x/net/html"
+)
+
+type Crawler struct {
+	visited      map[string]bool
+	brokenLinks  int
+	brokenReport []string
+	mu           sync.Mutex
+}
+
+func NewCrawler() *Crawler {
+	return &Crawler{
+		visited:      make(map[string]bool),
+		brokenReport: []string{},
+	}
+}
+
+func (c *Crawler) Visit(link string) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.visited[link] {
+		return false
+	}
+	c.visited[link] = true
+	return true
+}
+
+func (c *Crawler) ReportBrokenLink(link string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.brokenLinks++
+	c.brokenReport = append(c.brokenReport, link)
+}
+
+func (c *Crawler) Crawl(domain string, link string, depth int) {
+	if depth <= 0 {
+		return
+	}
+
+	if !c.Visit(link) {
+		return
+	}
+	fmt.Println("Visiting:", link)
+
+	parsedURL, err := url.Parse(link)
+	if err != nil {
+		fmt.Println("Error parsing URL:", link, err)
+		return
+	}
+
+	// Properly encode query parameters
+	parsedURL.RawQuery = parsedURL.Query().Encode()
+
+	encodedLink := parsedURL.String() // Fully encoded URL
+	fmt.Println("Encoded URL:", encodedLink)
+
+	resp, err := http.Get(encodedLink)
```

Comment:
Could we be potentially fixing issues be re-encoding the url rather than trying the raw link ?
```suggestion
	resp, err := http.Get(link)
```

Reply:
I have to encode it otherwise links with spaces like "Electronic monitoring" won't work

Reply:
it's all raw now